peekingduck/pipeline/nodes/input/visual.py

```diff
@@ -0,0 +1,257 @@
+from peekingduck.pipeline.nodes.node import AbstractNode
+
+
+class Node(AbstractNode):  # pylint: disable=too-many-instance-attributes
+    """Receives visual sources as inputs.
+
+    Inputs:
+        |none_input_data|
+
+    Outputs:
+        |img_data|
+
+        |pipeline_end_data|
+
+        |saved_video_fps_data|
+
+    Configs:
+        frames_log_freq (:obj:`int`): **default = 100**. |br|
+            Logs frequency of frames passed in CLI
+        mirror_image (:obj:`bool`): **default = False**. |br|
+            Flag to set extracted image frame as mirror image of input stream.
+        resize (:obj:`Dict`):
+            **default = { do_resizing: False, width: 1280, height: 720 }** |br|
+            Dimension of extracted image frame.
+        saved_video_fps (:obj:`int`): **default = 10**. |br|
+            FPS to be used to output the MP4 file after livestream is processed and
+            exported.  FPS is dependent on running machine performance.
+        source (:obj:`int`):
+            **default = https://storage.googleapis.com/peekingduck/videos/wave.mp4**. |br|
+            Input source can be: |br|
+            - filename : local image or video file |br|
+            - directory name : all media files will be processed |br|
+            - http URL for online cloud source : http[s]://... |br|
+            - rtsp URL for CCTV : rtsp://... |br|
+            - 0 for webcam live feed |br|
+            Refer to `OpenCV documentation
+            <https://docs.opencv.org/4.5.5/d8/dfe/classcv_1_1VideoCapture.html>`_
+            for more technical information.
+        threading (:obj:`bool`): **default = False**. |br|
+            Flag to enable threading when reading frames from camera.
+            The FPS can increase up to 30%.
+        buffer_frames (:obj:`bool`): **default = False**. |br|
+            Boolean to indicate if threaded class should buffer image frames.
+            If threading is True and output.media_writer is enabled, then
+            buffer_frames should be True to ensure output video is correctly
+            saved. One side effect of threading=True, buffer_frames=True is the
+            onscreen video display could appear laggy due to the buffering.
+    """
+
+    def __init__(self, config: Dict[str, Any] = None, **kwargs: Any) -> None:
+        super().__init__(config, node_path=__name__, **kwargs)
+        self._allowed_extensions = [
+            "gif",
+            "jpeg",
+            "jpg",
+            "png",
+            "avi",
+            "m4v",
+            "mkv",
+            "mov",
+            "mp4",
+        ]
+        self._file_name: str = ""
+        self._fps: float = 0  # self._fps > 0 if file playback
+        self.frame_counter: int = 0
+        self.tens_counter: int = 10
+        self.total_frame_count: int = 0
+        self.progress: int = 0
+        self.videocap: Union[None, VideoNoThread, VideoThread] = None
```

Comment:
I think we have been using `Optional` to type hint variables that can be `None` so this should probably be `Optional[Union[VideoNoThread, VideoThread]]` instead.

Using `Optional` can also potentially make things clearer with type aliasing [ref](https://stackoverflow.com/a/51710151)

Reply:
Changed.